Readme.md
# intercom.io
http://tarunc.github.io/intercom.io/

An API client in Node.JS for talking to intercom.io. This package implements the ~~complete API~~ (no longer; some of the API) for talking with the intercom.io API -- users, messages, impressions, tags, notes and more. This library's methods support promises (`methodName().then(resolve, reject)`)(thru [q](https://github.com/kriskowal/q)) and node style callbacks (`methodName(function(err, data){ /* callback code */ })`).

There complete docs can be found here - http://doc.intercom.io/api. See [docs](http://tarunc.github.io/intercom.io/) for complete API documentation of this library and all the methods that we support. See tests for more examples.

**NOTE: This library is not maintained be me anymore. Pull requests are still welcome and will be merged appropriately. I recommend using [Intercom's Node Library](https://github.com/intercom/intercom-node) instead.**

## Installation

To install the latest stable release with the command-line tool:
```sh
npm install --save intercom.io
```

## Usage

See [this library's docs](http://tarunc.github.io/intercom.io/) for complete API documentation and the [intercom API documentation](http://doc.intercom.io/api).

```javascript
var Intercom = require('intercom.io');

var options = {
  personalAccessToken: "your_access_token", // Intercom Access Token
  timeout: 60 * 1000 // Timeout (in ms) for requests to the intercom API
};

var intercom = new Intercom(options);
// Can also be written as:
// var intercom = new Intercom("your_access_token");


// Usage with appId and apiKey
// appId and appKey will be disabled on intercom as of Jan 2017 and will be replaced by Personal Access Tokens
// See https://developers.intercom.com/docs/personal-access-tokens
var options = {
  apiKey: "your_API_key", // Your Personal Intercom API Key
  appId: "your_APP_ID" // Your Intercom App Id
};

var intercom = new Intercom(options);
// Can also be written as:
// var intercom = new Intercom("your_APP_ID", "your_API_key");

// Note: you can also require and create an instance in the same step if you would like.
// Example:
// var intercom = require('intercom.io').create("your_APP_ID", "your_API_key");
// or
// var intercom = require('intercom.io').create(options);

// To create a user
// Every method supports promises or callbacks.
intercom.createUser({
  "email" : "ben@intercom.io",
  "user_id" : "7902",
  "name" : "Ben McRedmond",
  "created_at" : 1257553080,
  "custom_data" : {"plan" : "pro"},
  "last_seen_ip" : "1.2.3.4",
  "last_seen_user_agent" : "ie6",
  "companies" : [
    {
      "id" : 6,
      "name" : "Intercom",
      "created_at" : 103201,
      "plan" : "Messaging",
      "monthly_spend" : 50
    }
  ],
  "last_request_at" : 1300000000
}, function(err, res) {
  // err is an error object if there was an error
  // res is **JSON** response
  // In this case:
  // {
  //   "intercom_id": "52322b3b5d2dd84f23000169",
  //   "email": "ben@intercom.io",
  //   "user_id": "7902",
  //   "name": "Ben McRedmond",
  //   "created_at": 1257553080,
  //   "last_impression_at": 1300000000,
  //   "custom_data": {
  //     "plan": "pro"
  //   },
  // ...
  // ...
  // ...
  //   "session_count": 0,
  //   "last_seen_ip": "1.2.3.4",
  //   "last_seen_user_agent": "ie6",
  //   "unsubscribed_from_emails": false
  // }
});

// To get a user
// (using a promise)
intercom.getUser({ "email": "ben@intercom.io" }).then(function(res) {
  // res is **JSON** response
  // In this case:
  // {
  //   "intercom_id": "52322b396823b17b1100016a",
  //   "email": "ben@intercom.io",
  //   "user_id": "7902",
  //   "name": "Ben McRedmond",
  //   "created_at": 1257553080,
  //   "last_impression_at": 1300000000,
  //   "custom_data": {
  //     "plan": "pro"
  //   },
  // ...
  // ...
  // ...
  //   "session_count": 0,
  //   "last_seen_ip": "1.2.3.4",
  //   "last_seen_user_agent": "ie6",
  //   "unsubscribed_from_emails": false
  // }
}, function(err) {
  // err is an error object if there was an error
});

// To get multiple users
intercom.getUsers({
  page: 1,
  per_page: 500,
  tag_id: 7002,
  tag_name: "me"
}, function (err, res) {
  // err is an error object
  // res is the **JSON** response
  // {
  //   "users": [
  //     {
  //       "intercom_id": "52322b366823b173eb000170",
  //       "email": "first.user@example.com",
  //       "user_id": "123",
  //       "name": "First User",
  //       "created_at": 1270000000,
  //       "last_impression_at": 1300000000,
  //       "custom_data": {
  //         "app_name": "Genesis",
  // ...
  // ...
  // ...
  //       "last_seen_user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_3) AppleWebKit/535.11 (KHTML, like Gecko) Chrome/17.0.963.56 Safari/535.11",
  //       "unsubscribed_from_emails": false
  //     }
  //   ],
  //   "total_count": 3,
  //   "page": 1,
  //   "next_page": null,
  //   "previous_page": null,
  //   "total_pages": 1
  // }
});
```

List of supported methods: (More methods on [docs](http://tarunc.github.io/intercom.io/).)
```javascript
* intercom.getPages
* intercom.getUsers
* intercom.getUser
* intercom.viewUser
* intercom.createUser
* intercom.updateUser
* intercom.deleteUser
* intercom.bulkAddUsers
* intercom.createContact
* intercom.updateContact
* intercom.deleteContact
* intercom.getContact
* intercom.viewContact
* intercom.getContacts
* intercom.convertContact
* intercom.listCompanies
* intercom.viewCompany
* intercom.createCompany
* intercom.updateCompany
* intercom.listCompanyUsers
* intercom.listAdmins
* intercom.createNote
* intercom.listNotes
* intercom.viewNote
* intercom.getTag
* intercom.createTag
* intercom.deleteTag
* intercom.updateTag
* intercom.listSegments
* intercom.viewSegment
* intercom.createEvent
* intercom.getCounts
* intercom.createUserMessage
* intercom.listConversations
* intercom.getConversation
* intercom.replyConversation
* intercom.markConversationAsRead
* intercom.closeConversation
```

See [docs](http://tarunc.github.io/intercom.io/) for complete API documentation and the [intercom API documentation](http://doc.intercom.io/api). See tests for more examples.

__Note__: Every method returns a promise but accepts callbacks too.

## License

(The MIT License)

Copyright (c) 2013-2018 Tarun Chaudhry &lt;opensource@chaudhry.co&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
1.5.0 / 2018-01-02
==================

  * Store Error Details from Intercom into `IntercomError`
  * Fix `getPages` only returns single page if no more pages

1.4.0 / 2016-10-28
==================

  * Removed requirement for `apiKey`
  * Updated support for Personal Access Tokens
  * Added support for a request timeout

1.3.0 / 2016-10-22
==================

  * Updated `debug` dependency
  * Updated `request` dependency
  * Changed path for `bulkAddUsers`
  * Extended result with intercom's ratelimit information
  * Changed to use caret version ranges, not tilde in package.json
  * Improved `getPages`
  * Added support for Personal Access Tokens

1.2.1 / 2016-02-15
==================

  * Fixed `markConversationAsRead`

1.2.0 / 2016-02-12
==================

  * Added methods to support conversations.

1.1.0 / 2016-02-11
==================

  * Add contact methods

1.0.10 / 2015-06-05
==================

  * Update `deleteUser`

1.0.9 / 2015-05-27
==================

  * Update `deleteTag` and `updateTag`

1.0.8 / 2015-04-25
==================

  * Added `intercom.getPages` support

1.0.7 / 2015-03-06
==================

  * Added JSHINT Support to Gruntfile
  * Updated `lodash` usage

1.0.5 / 2014-10-24
==================

  * Update `request` dependency

1.0.4 / 2014-10-24
==================

  * Update `qs` dependency

1.0.3 / 2014-09-20
==================

  * Added support for Conversations

1.0.2 / 2014-09-20
==================

  * User POST for updateUser

1.0.1 / 2014-09-12
==================

  * Support for intercom.io errors

1.0.0 / 2014-09-11
==================

  * Major overhaul to the API

0.0.8 / 2014-05-04
==================

  * Update support for promises

0.0.7 / 2014-05-04
==================

  * Added support for promises

0.0.6 / 2014-04-27
==================

  * Added createEvent

0.0.5 / 2014-02-18
==================

  * Fixed delete user

0.0.4 / 2013-10-31
==================

  * Fixed JSON Body Errors

0.0.3 / 2013-10-05
==================

  * Added basic integration tests

0.0.2 / 2013-10-05
==================

  * Added docs
  * Made it work with just passing in options

0.0.1 / 2013-07-03
==================

  * Initial release


test/test.js
var assert = require('assert');

// Please set up the approprite environment variables
var APP_ID = process.env.INTERCOM_APP_ID;
var API_KEY = process.env.INTERCOM_API_KEY;

var PERSONAL_ACCESS_TOKEN = process.env.INTERCOM_ACCESS_TOKEN;

var intercom = require('../index.js').create(APP_ID, API_KEY);
// events api requires v3 endpoint
var intercom_v3 = require('../index.js').create(APP_ID, API_KEY, {
  endpoint: 'https://api.intercom.io/'
});

// Tests with a PERSONAL_ACCESS_TOKEN
// var intercom = require('../index.js').create(PERSONAL_ACCESS_TOKEN);
// // events api requires v3 endpoint
// var intercom_v3 = require('../index.js').create({
//   personalAccessToken: PERSONAL_ACCESS_TOKEN,
//   endpoint: 'https://api.intercom.io/'
// });


describe('Intercom', function() {
  describe('#getUsers()', function() {
    it('should get all users', function(done) {
      intercom.getUsers(function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#createUser()', function() {
    it('should create a specific user', function(done) {
      intercom.createUser({
        'email' : 'somebody@example.com',
        'name' : 'Somebody',
        'created_at' : seconds_since_epoch(),

        'last_seen_ip' : '1.2.3.4',
        'last_seen_user_agent' : 'ie6'
      }, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#updateUser()', function() {
    it('should update a specific user', function(done) {
      intercom.updateUser({
        'email' : 'somebody@example.com',
        'name' : 'Me!'
      }, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#getUser()', function() {
    it('should get a specific user', function(done) {
      intercom.getUser({ email: 'somebody@example.com' }, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#createEvent()', function() {
    it('should create a user event', function(done) {
      intercom_v3.createEvent({
        'event_name' : 'some event',
        'created' : seconds_since_epoch(),
        'email' : 'somebody@example.com'
      }, function (err, res) {
        if (err) {
            console.error(err);
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#deleteUser()', function() {
    it('should delete a specific user', function(done) {
      intercom.deleteUser({
        'email' : 'somebody@example.com'
      }, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#listConversations()', function() {
    it('should list out conversations', function(done) {
      intercom.listConversations({}, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#getConversation()', function() {
    it('should show a specific conversation', function(done) {
      intercom.getConversation({ id: 858210458 }, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });

  describe('#getPages()', function() {
    it('should get all the pages of an Intercom resource in parallel for events.', function(done) {
      this.timeout(20000);
      intercom.getPages('events', { user_id: '101', type: 'user'}, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });

    it('should get all the pages of an Intercom resource in parallel for users.', function(done) {
      this.timeout(20000);
      intercom.getPages('users', {}, function (err, res) {
        if (err) {
          throw err;
        }

        console.log(res);
        done();
      });
    });
  });
});


function seconds_since_epoch(){ return Math.floor( Date.now() / 1000 ) }


test/contacts.spec.js
// Please set up the approprite environment variables
var APP_ID = process.env.INTERCOM_APP_ID;
var API_KEY = process.env.INTERCOM_API_KEY;

var PERSONAL_ACCESS_TOKEN = process.env.INTERCOM_ACCESS_TOKEN;

var intercom = require('../index.js').create(APP_ID, API_KEY);

// To test with a PERSONAL_ACCESS_TOKEN, uncomment line below
// var intercom = require('../index.js').create(PERSONAL_ACCESS_TOKEN);

// Chai Setup
// -------------------------
global.chai   = require('chai');
global.expect = chai.expect;
global.assert = chai.assert;

var chaiSubset = require('chai-subset');
chai.use(chaiSubset);

// Chai Promise Setup
var chaiAsPromised = require('chai-as-promised');
    chai.use(chaiAsPromised);

// Sinon Setup
global.sinon = require('sinon');
var sinonChai = require('sinon-chai');


global.should = chai.should();
chai.use(sinonChai);

// Replay (Like Ruby's VCR)
global.Replay  = require('replay');
Replay.fixtures = __dirname + '/fixtures/replay';

// Set to 'record' to record new callouts
Replay.mode = 'replay';



describe('Contacts', function() {
  var my_contact;
  var contact = {
    name: 'Ray Velcoro',
    email: 'ray@hbo.com'
  };

  before(function(done) {
    intercom.updateContact(contact).then(function(obj) {
      my_contact = obj;
      done();
    })
  });

  describe('#createContact()', function() {

    it('should create and return a contact', function() {
      return intercom.createContact(contact).should.be.fulfilled
      .then(function(obj){
        expect(obj).to.containSubset(contact);
      });
    });

  });

  describe('#updateContact()', function() {

    it('should create and return a contact', function() {
      return intercom.updateContact(contact).should.be.fulfilled
      .then(function(obj){
        expect(obj).to.containSubset(contact);
      });
    });

  });

  describe('#getContact()', function() {

    it('should return our contact', function() {
      expect(my_contact).to.containSubset(contact);
    });

  });


  // Currently not working as expected
  // describe('#deleteContact()', function() {
  //
  //   it('should return our contact', function() {
  //     return intercom.deleteContact(my_contact).should.be.fulfilled
  //     .then(function(obj){
  //       expect(obj).to.containSubset(my_contact);
  //     });
  //   });
  //
  // });

  describe('#convertContact()', function() {

    var contact = {
      name: 'Frank Semyon',
      email: 'frank@hbo.com'
    };

    it('should convert the contact', function() {

      return intercom.createContact(contact).should.be.fulfilled.then(function(contact){
        return intercom.convertContact({
          contact: { user_id: contact.user_id },
          user: { email: contact.email }
        }).should.be.fulfilled

      }).catch(function(e) { console.log(e)})

      .then(function(user){
        user.type.should.equal('user');

        // Successfully converted a user
        // Clean it up
        return intercom.deleteUser(user).should.be.fulfilled;
      });

    });

  });


});


Gruntfile.js
/*
 * Gruntfile.js
 */

'use strict';

// intercom.io uses `[grunt](http://gruntjs.com)` to define and run repetitive tasks
// such as minification, compilation, unit testing, linting, etc.
var path = require('path');

// ## Grunt configuration.
// @export `ConfigureGruntService` a function which configures tasks to be run with
// `grunt`
module.exports = function ConfigureGruntService(grunt) {
  // load all grunt tasks
  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

  // Provide some basic configuration for grunt
  grunt.initConfig({
    // Read in the package.json to get access to the variables defined in there
    pkg: grunt.file.readJSON('package.json'),
    // Define the Docker task to generate documentation from our codebase.
    docker: {
      app: {
        expand: true,
        src: ['lib/*.js'],
        dest: 'docs/',
        options: {
          onlyUpdated: false,
          colourScheme: 'default',
          ignoreHidden: false,
          sidebarState: false,
          exclude: false,
          lineNums: true,
          multiLineOnly: false,
          extras: ['fileSearch', 'goToLine']
        }
      }
    },
    // Define the various shell tasks
    shell: {
      // Define a `sweetenDocker` command which sweetens the output from docker
      sweetenDocker: {
        command: 'bin/sweeten-docker ./docs/lib',
        stdout: true,
        stderr: true,
        failOnError: true
      }
    },
    jshint: {
      options: {
        jshintrc: true
      },
      files: {
        src: ['lib/*.js', 'index.js', 'Gruntfile.js']
      }
    },
    'gh-pages': {
      options: {
        base: 'docs',
        push: true,
        message: 'Auto-generated commit'
      },
      src: '**/*'
    },
    copy: {
      docs: {
        files: [{
          src: 'docs/intercom.io.js.html',
          dest: 'docs/index.html'
        }]
      }
    },
    'string-replace': {
      docs: {
        files: [{
          expand: true,
          cwd: 'docs/lib/',
          src: '*',
          dest: 'docs/'
        }],
        options: {
          replacements: [{
            pattern: /\.\.\//g,
            replacement: './'
          }]
        }
      }
    }
  });

  // ## Define the tasks
  // ### Default task(s).
  // Just running `grunt` will load the default tasks.
  // For now, the only default tasks are `jshint:app`
  grunt.registerTask('default', ['jshint']);

  // ### Doc task(s).
  // Running `grunt doc` will generate documentation for intercom.io.
  // The documentation will be put into the `./docs` folder in project's root
  // folder.
  grunt.registerTask('doc', ['docker:app', 'shell:sweetenDocker', 'string-replace:docs', 'copy:docs']);
  // @note `grunt docs` provides an alias for the `grunt doc` task
  grunt.registerTask('docs', ['doc']);
};


lib/intercom.io.js
/**
 * Module dependencies.
 */
var request = require('request'),
    qs = require('qs'),
    _ = require('lodash'),
    debug = require('debug')('intercom.io'),
    // Represents errors thrown by intercom, see [IntercomError.js](./IntercomError.js.html)
    IntercomError = require('./IntercomError'),
    url = require('url'),
    Q = require('q');

// Define some sane default options
// Right now we only have one endpoint for intercom data
var defaultOptions = {
  // Intercom's API endpoint to hit
  endpoint: 'https://api.intercom.io/',

  // Add a default imeout (in ms) of a minute
  timeout: 60 * 1000
};

/**
 * `Intercom` constructor.
 *
 * @param {String} appId - your app Id OR personalAccessToken
 * @param {String} apiKey - your api key
 * @param {Object} options - an options object
 * @api public
 */
function Intercom(appId, apiKey, options) {
  // Overload the contractor
  // Parse out single option objects
  if (_.isObject(appId) && !_.isString(appId) && ((appId.apiKey && appId.appId) || appId.personalAccessToken)) {
    apiKey = appId.apiKey || '';
    options = _.omit(appId, 'apiKey', 'appId');
    appId = appId.appId || appId.personalAccessToken;
  }

  // Preform some sane validation checks and throw errors
  // We need the appId
  if (!appId) {
    throw new IntercomError('Invalid App ID: ' + appId);
  }

  // Copy over the relavant data
  this.appId = appId;
  this.apiKey = apiKey || '';

  // Extend the defaults
  this.options = _.defaults(options || {}, Intercom.defaultOptions);

  // Contruct the endpoint with the correct auth from the appId and apiKey
  this.endpoint = this.options.endpoint;
}

/**
 * Expose `defaultOptions` for the intercom library so that this is changable.
 */
Intercom.defaultOptions = defaultOptions;

/**
 * Helper method to create an instance easily
 *
 * Enables use like this:
 *
 *     `var intercom = require('intercom.io').create("your_APP_ID", "your_API_key");`
 *
 *      or
 *
 *     `var intercom = require('intercom.io').create(options);`
 *
 * @param {String} appId - your app Id
 * @param {String} apiKey - your api key
 * @param {Object} options - an options object
 * @api public
 */
Intercom.create = function(appId, apiKey, options) {
  return (new Intercom(appId, apiKey, options))
};

/**
 * The main method that makes all the requests to intercom.
 * This method deals with the intercom api and can be used
 * to make requests to the intercom api.
 *
 * @api public
 */
Intercom.prototype.request = function(method, path, parameters, cb) {
  debug('Requesting [%s] %s with data %o', method, this.endpoint + path, parameters);

  var url = _.includes(path, this.endpoint) ? path : this.endpoint + path;

  var requestOptions = {
    method: method,
    url: url,
    timeout: this.options.timeout
  };

  if (method === 'GET') {
    requestOptions.qs = parameters;
    requestOptions.headers = {
      'Accept': 'application/json'
    };
  } else {
    // Intercom.io now requires parameters in JSON
    // requestOptions.form = parameters;
    requestOptions.body = JSON.stringify(parameters);
    requestOptions.headers = {
      'Accept': 'application/json',
      'Content-Type' : 'application/json'
    };
  }

  requestOptions.auth = {
    username: this.appId,
    password: this.apiKey || ''
  };

  // create a promise to return
  var deferred = Q.defer();

  request(requestOptions, function(err, res, data) {
    if (err) {
      // Reject the promise
      return deferred.reject(err);
    }

    // Try to parse the data
    var parsed;
    if (data) {
      debug('Received response %s', data);

      try {
        parsed = JSON.parse(data);

        if (parsed && (parsed.error || parsed.errors)) {
          var errorCodes = '"' + parsed.errors.map(function (error) {
              return error.code;
            }).join('", "') + '"';

          // Reject the promise
          return deferred.reject(new IntercomError(errorCodes + ' error(s) from Intercom', parsed.errors));
        }
      } catch (exception) {
        // Reject the promise
        // return deferred.reject(exception);
      }
    }

    var headers = res.headers;
    data = parsed || data;
    data.meta = {
      ratelimit_limit: headers['x-ratelimit-limit'],
      ratelimit_remaining: headers['x-ratelimit-remaining'],
      ratelimit_reset: headers['x-ratelimit-reset']
    };

    // Resolve the promise
    return deferred.resolve(data);
  });

  // Return the promise and promisify any callback provided
  return deferred.promise.nodeify(cb);
};

/**
 * GETs all the pages of an Intercom resource in parallel.
 * @param {String} path The resource to retrieve (e.g. companies)
 * @param {Object} parameters Query parameters for the root resource
 * @param {Function} cb Optional request callback
 * @returns {Promise} A promise of an array containing all the elements of the
 * requested resource.
 *
 * @api public
 */
Intercom.prototype.getPages = function(path, parameters, cb) {
  // TODO Add a progress indicator when each page is retrieved
  var self = this;

  var rootPromise = this.request('GET', path, parameters);

  function requestNextPage(pages, nextUrl, parameters) {
    return self.request('GET', nextUrl, parameters).then(function(root) {
      var next = _.get(root, 'pages.next');
      if(next) {
        pages.push(root);
        return requestNextPage(pages, next, parameters);
      }
      return pages.concat(root);
    });
  }

  function getAllPages(root) {
    var next = _.get(root, 'pages.next');
    if (next) {
      return requestNextPage([root], next, parameters);
    }
    return [root];
  }

  function gatherPages (acc, result) {
    return acc.concat(result[path]);
  }

  return rootPromise.then(function (root) {
    return root;
  }).then(function (root) {
    return getAllPages(root);
  }).then(function (results) {
    return results.reduce(gatherPages, []);
  }).nodeify(cb);
};

/**
 * Helper method to create dates for intercom easily
 *
 * @param {Number  or Date} date - the date (in milliseconds if Number). Leave empty if you want the date set to now
 * @returns {Number} epoch - time in seconds since Epoch (how intercom handles dates)
 * @api public
 */
Intercom.prototype.date = function(date) {
  if (date === undefined) {
    date = new Date();
  }

  if (date && _.isDate(date) && date.getTime) {
    return Math.floor(date.getTime() / 1000);
  }

  return Math.floor((new Date(date)) / 1000);
};

// ### Users
Intercom.prototype.getUsers = function(options, cb) {
  if (_.isFunction(options)) {
    cb = options;
    options = {};
  }

  return this.request('GET', 'users', options, cb);
};

Intercom.prototype.getUser = function(userObj, cb) {
  return this.request('GET', 'users', userObj, cb);
};

Intercom.prototype.viewUser = function(userObj, cb) {
  if (userObj && userObj.hasOwnProperty('id')) {
      return this.request('GET', 'users' + '/' + userObj.id, {}, cb);
  } else {
      return this.request('GET', 'users', userObj, cb);
  }
};

Intercom.prototype.createUser = function(userObj, cb) {
  return this.request('POST', 'users', userObj, cb);
};

Intercom.prototype.updateUser = function(userObj, cb) {
  return this.request('POST', 'users', userObj, cb);
};

Intercom.prototype.deleteUser = function(userObj, cb) {
  if (userObj && userObj.hasOwnProperty('id')) {
    return this.request('DELETE', 'users' + '/' + userObj.id, {}, cb);
  }
  else {
    return this.request('DELETE', 'users', userObj, cb);
  }
};

Intercom.prototype.bulkAddUsers = function(userObj, cb) {
  return this.request('POST', 'bulk/users', userObj, cb);
};


// ### Contacts
Intercom.prototype.createContact = function(obj, cb) {
  return this.request('POST', 'contacts', obj, cb);
};
Intercom.prototype.updateContact = Intercom.prototype.createContact;

Intercom.prototype.deleteContact = function(obj, cb) {
  return this.request('DELETE', 'contacts', obj, cb);
};

Intercom.prototype.getContact = function(options, cb) {
  if (_.isFunction(options)) {
    cb = options;
    options = {};
  }

  return this.request('GET', 'contacts', options, cb);
};
Intercom.prototype.getContacts = Intercom.prototype.viewContact = Intercom.prototype.getContact;

Intercom.prototype.convertContact = function(obj, cb) {
  return this.request('POST', 'contacts/convert', obj, cb);
};

// ### Companies
Intercom.prototype.listCompanies = function(companyObj, cb) {
  return this.request('GET', 'companies', companyObj, cb);
};

Intercom.prototype.viewCompany = function(companyObj, cb) {
  if (companyObj && companyObj.hasOwnProperty('id')) {
      return this.request('GET', 'companies/' + companyObj.id, {}, cb);
  } else {
      return this.request('GET', 'companies', companyObj, cb);
  }
};

Intercom.prototype.createCompany = function(companyObj, cb) {
  return this.request('POST', 'companies', companyObj, cb);
};

Intercom.prototype.updateCompany = function(companyObj, cb) {
  return this.request('POST', 'companies', companyObj, cb);
};

/*
// Functionality doesn't currently exist
Intercom.prototype.deleteCompany = function(companyObj, cb) {
  // return this.request('DELETE', 'companies', companyObj, cb);
};
*/

Intercom.prototype.listCompanyUsers = function(companyObj, cb) {
  if (companyObj && companyObj.hasOwnProperty('id')) {
      return this.request('GET', 'companies/' + companyObj.id + '/users', {}, cb);
  } else {
      return this.request('GET', 'companies', companyObj, cb);
  }
};

// ### Admins
Intercom.prototype.listAdmins = function(adminObj, cb) {
  return this.request('GET', 'admins', adminObj, cb);
};

// ### Notes
Intercom.prototype.createNote = function(noteObj, cb) {
  return this.request('POST', 'notes', noteObj, cb);
};

Intercom.prototype.listNotes = function(noteObj, cb) {
  return this.request('GET', 'notes', noteObj, cb);
};

Intercom.prototype.viewNote = function(noteObj, cb) {
  if (noteObj && noteObj.hasOwnProperty('id')) {
    return this.request('GET', 'notes/' + noteObj.id, {}, cb);
  } else {
    return this.request('GET', 'notes', noteObj, cb);
  }
};

// ### Tags
Intercom.prototype.getTag = function(tagObj, cb) {
  return this.request('GET', 'tags', tagObj, cb);
};

Intercom.prototype.createTag = function(tagObj, cb) {
  return this.request('POST', 'tags', tagObj, cb);
};

Intercom.prototype.deleteTag = function(tagObj, cb) {
  return this.request('DELETE', 'tags/' + tagObj.id, {}, cb);
};

Intercom.prototype.updateTag = function(tagObj, cb) {
  // Note: tagObj should be empty
  return this.request('POST', 'tags', tagObj, cb);
};

// ### Segments
Intercom.prototype.listSegments = function(segmentObj, cb) {
  return this.request('GET', 'segments', segmentObj, cb);
};

Intercom.prototype.viewSegment = function(segmentObj, cb) {
  if (segmentObj && segmentObj.hasOwnProperty('id')) {
    return this.request('GET', 'segments/' + segmentObj.id || segmentObj.segment_id, {}, cb);
  } else {
    return this.request('GET', 'segments', segmentObj, cb);
  }
};

// ### Events
Intercom.prototype.createEvent = function(eventObj, cb) {
  return this.request('POST', 'events', eventObj, cb);
};

// ### Counts
Intercom.prototype.getCounts = function(countObj, cb) {
  return this.request('GET', 'counts', countObj, cb);
};

// ### Conversations
// (NOTE: the Conversations API is only available to Apps on the Starter and Premium plans)
Intercom.prototype.createUserMessage = function(messageObj, cb) {
  return this.request('POST', 'messages', messageObj, cb);
};

Intercom.prototype.listConversations = function (convObj, cb) {
  return this.request('GET', 'conversations', convObj, cb);
};

Intercom.prototype.getConversation = function (convObj, cb) {
  if (convObj && convObj.hasOwnProperty('id')) {
    return this.request('GET', 'conversations/' + convObj.id, {}, cb);
  } else {
    return this.request('GET', 'conversations', convObj, cb);
  }
};

// It will automatically reply to the last conversation if `convObj.id` is left out
Intercom.prototype.replyConversation = function (convObj, cb) {
  if (convObj && convObj.hasOwnProperty('id')) {
    return this.request('POST', 'conversations/' + convObj.id + '/reply', convObj, cb);
  } else {
    return this.request('POST', 'conversations/last/reply', convObj, cb);
  }
};
// Closing a conversation is the same as replying to a conversation with a special `convObj`
// See https://developers.intercom.io/docs/closing-a-conversation
Intercom.prototype.closeConversation = Intercom.prototype.replyConversation;

Intercom.prototype.markConversationAsRead = function (convObj, cb) {
  if (convObj && convObj.hasOwnProperty('id')) {
    return this.request('PUT', 'conversations/' + convObj.id , convObj, cb);
  } else {
    return this.request('PUT', 'conversations/last', convObj, cb);
  }
};

/**
 * Expose `Intercom` Library.
 */
module.exports = Intercom;
